plugins/FakeCardReader.cpp

```diff
@@ -198,70 +197,68 @@ FakeCardReader::generate_data(appfwk::DAQSink<std::unique_ptr<types::WIB_SUPERCH
 }
 
 void 
-FakeCardReader::generate_tp_data(appfwk::DAQSink<std::unique_ptr<types::TP_SUPERCHUNK_STRUCT>>* myqueue, int linkid) 
+FakeCardReader::generate_tp_data(appfwk::DAQSink<std::unique_ptr<types::RAW_WIB_TP_STRUCT>>* myqueue, int linkid) 
 {
   std::stringstream ss;
   ss << "card-reader-" << linkid;
   pthread_setname_np(pthread_self(), ss.str().c_str());
   // Init ratelimiter, element offset and source buffer ref
-  dunedaq::readout::RateLimiter rate_limiter(cfg_.rate_khz);
+  dunedaq::readout::RateLimiter rate_limiter(cfg_.tp_rate_khz);
   rate_limiter.init();
   int offset = 0;
   auto& source = tp_source_buffer_->get();
 
   // This should be changed in case of a generic Fake ELink reader (exercise with TPs dumps)
   int num_elem = tp_source_buffer_->num_elements();
-  uint64_t ts_0 = reinterpret_cast<dunedaq::dataformats::TPFrame*>(source.data())->tp_header()->timestamp();
+  uint64_t ts_0 = reinterpret_cast<dunedaq::dataformats::RawWibTp*>(source.data())->get_header()->timestamp();
   ERS_INFO("First timestamp in the source file: " << ts_0 << "; linkid is: " << linkid);
   uint64_t ts_next = ts_0;
 
-  struct TPSubframe {
-    word_t w1;
-    word_t w2;
-    word_t w3;
-  };
-  bool head = false;
-  dunedaq::dataformats::TPFrame* tf = nullptr;
-  const dunedaq::dataformats::TPDataBlock* tb = nullptr;
+  //dunedaq::dataformats::RawWibTp* tf{nullptr};
+  std::unique_ptr<dunedaq::dataformats::RawWibTp> tf{nullptr};
+  const dunedaq::dataformats::TpDataBlock* tb{nullptr};
   while (run_marker_.load()) {
       // Which element to push to the buffer
-      if (offset == num_elem) {
+      if (offset == num_elem * static_cast<int>(constant::RAW_WIB_TP_SUBFRAME_SIZE)) {
         offset = 0;
-      } 
-      // Create next superchunk
-      std::unique_ptr<types::TP_SUPERCHUNK_STRUCT> payload_ptr = std::make_unique<types::TP_SUPERCHUNK_STRUCT>();
-      // Memcpy from file buffer to flat char array
-      ::memcpy((void*)&payload_ptr->data, (void*)(source.data()+offset*constant::TP_SUPERCHUNK_SIZE), constant::TP_SUPERCHUNK_SIZE);
-
-      for (unsigned int i=0; i<constant::TP_SUPERCHUNK_SIZE/12; ++i) {
-        auto* sp = reinterpret_cast<TPSubframe*>(((uint8_t*)payload_ptr.get())+i*12);
-        if (!head) {
-          tf = new dunedaq::dataformats::TPFrame();
-          tb = tf->tp_data();
-          const dunedaq::dataformats::TPHeader* tfh = tf->tp_header();
-          tfh = reinterpret_cast<dunedaq::dataformats::TPHeader*>(sp);
-          tf->set_tp_header(*tfh);
+      }
+      // Count number of subframes in a TP frame
+      int n = 1;
+      while (reinterpret_cast<types::TpSubframe*>(((uint8_t*)source.data()) 
+             + offset + (n-1)*constant::RAW_WIB_TP_SUBFRAME_SIZE)->word3 != 0xDEADBEEF) {
+        n++;
+      }
+      // Create next TP frame
+      std::unique_ptr<types::RAW_WIB_TP_STRUCT> payload_ptr = std::make_unique<types::RAW_WIB_TP_STRUCT>();
+      m_data_ = malloc(n*constant::RAW_WIB_TP_SUBFRAME_SIZE);
```

Comment:
I'm really not keen on calling malloc. We will need to have a chat what is the mechanism here and if there are potential improvements.